_posts/2022-04-01-First-steps-with-swift-concurrency.md

```diff
@@ -0,0 +1,341 @@
+
+```swift
+public actor ThumbnailDataSource { }
+```
+
+Great, now we define the function which will start fetching the thumbnails.
+
+```swift
+public actor ThumbnailDataSource { 
+	public typealias ThumbnailsStream = AsyncStream<(String, UIImage)>
+
+	public nonisolated func getThumbnails(models: [ApodModel]) -> ThumbnailsStream {
+		return ThumbnailsStream { continuation in
+			Task {
+				try await fetchThumbnails(from: models, continuation: continuation)
+			}
+		}
+	}
+}
+```
+
+As a parameter we take the array of ApodModels which contains the url for our images. Our function will return the `AsyncStream`, which will be delivering the tuple `(String, UIImage)` when the particular image is fetched (`String` will contain the image URL).
+
+Now we create the `AsyncStream` using its intializer, which is returning the `continuation` as a closure parameter, which we will use later on (more about [AsyncStream](https://developer.apple.com/documentation/swift/asyncstream)). In short continuation is needed to inform the stream about the received elements or when the stream is finishing its work.
+
+Please note that `func getThumbnails()` is not an `async` function, because we would like to only bind the data and receive them later on. That's why all async code was wrapped as a `Task`. The `nonisolated` prefix gave a possibility for the function to synchronously create the `AsyncStream` in the code. Since we do not modify anything in terms of the actors properties in the function, it is safe to expose the method as `nonisolated`.
+
+And then the usage of the Task Group looks like below:
+
+```swift
+private func fetchThumbnails(
+        from apody: [ApodModel],
+        continuation: ThumbnailsStream.Continuation
+    ) async throws {
+        try await withThrowingTaskGroup(of: (String, UIImage).self) { group in
+            for apod in apody.filter({ $0.media_type == .image }) {
+                group.addTask(priority: .background) {
+                    (apod.url, try await self.fetchImage(url: apod.url))
+                }
+            }
+            for try await data in group {
+                continuation.yield(data)
+            }
+            continuation.finish()
+        }
+}
+```
+
+Once added a child task to a group (by calling `group.addTask()`), child tasks are being executed immediately and in any order.
+
+The `continuation` property yields the result every time the group task finishes its fetch. Once all of the data are fetched, we call `conitnuation.finish()` to inform that the stream has finished publishing data.
+
+The binding in the parent would look like this:
+
+```swift
+@MainActor class ApodViewModel: ObservableObject {
+ 
+	@Published var thumbnails: [String: UIImage] = [:]
+
+	private let dataSource: ThumbnailDataSource
+	...
+	public func fetchThumbnails(for models: [ApodModel]) async {
```

Comment:
Is the `async` keyword required when `await` is used below? I read this code as synchronous (my understanding is that `await` waits for the values returned by the stream).

Reply:
Yes, when function uses the `await` call in its body, we have to:
a) Mark the function as `async` to tell the compiler that this is an `async` function.
b) wrap the `await` part of the code into `Task { for await ... }`
Otherwise Xcode will yield an error and won't allow to compile the code.